GoogleDataTransport/GDTCORLibrary/GDTCORPlatform.m

```diff
@@ -330,21 +330,100 @@ - (instancetype)init {
   return self;
 }
 
+/** Creates a NSMutableDictionary for mapping bgID to task semaphore.
+ *
+ * @return A NSMutableDictionary singleton instance for mapping bgID to task semaphore.
+ */
++ (NSMutableDictionary<NSNumber *, dispatch_semaphore_t> *)getBgIDToSemaphoreMap {
+  static NSMutableDictionary<NSNumber *, dispatch_semaphore_t> *sBgIDToSemephoreMap;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    sBgIDToSemephoreMap = [[NSMutableDictionary alloc] init];
+  });
+  return sBgIDToSemephoreMap;
+}
+
+/** Creates a queue on which all task semaphore will populate and remove from map.
+ *
+ * @return A dispatch queue singleton instance for the lifetime of an application.
+ */
++ (dispatch_queue_t)getSemaphoreQueue {
+  static dispatch_queue_t sSemaphoreQueue;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    sSemaphoreQueue = dispatch_queue_create("com.google.GDTCORApplication", DISPATCH_QUEUE_SERIAL);
+  });
+  return sSemaphoreQueue;
+}
+
+/** Populates a unique NSUInteger as GDTCORBackgroundIdentifier for given semaphore and maps it to
+ * the semaphore in NSMutableDictionary.
+ *
+ * @param semaphore The task semaphore needs to be mapping.
+ * @return A unique GDTCORBackgroundIdentifier mapped to the given semaphore.
+ */
++ (GDTCORBackgroundIdentifier)populateBgIDwithSemaphore:(dispatch_semaphore_t)semaphore {
+  __block GDTCORBackgroundIdentifier bgID = GDTCORBackgroundIdentifierInvalid;
+  if ([self getSemaphoreQueue]) {
```

Comment:
You should assign this to a variable for 2 reasons:
1. To reduce the number of times you send a message (call a method)
2. Without a strong reference to a variable, there's no guarantee that the object from the first method call is the same as the second, or that the 2nd time you call the method that it'll return the same thing.

Learning more about strong and weak variables could be useful. Here's one link to help learn more: https://stackoverflow.com/a/11013715

Reply:
Changed queue to a class wide static variable.